In [2]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [3]:
# Load the given datasets
numerical_data = pd.read_csv("/Users/<redacted>/Documents/CS383/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_numerical.csv")
categorical_data = pd.read_csv("/Users/<redacted>/Documents/CS383/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv")
# Print the data

numerical_data,categorical_data

(     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
 0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
 1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
 2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
 3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
 4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
 ..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
 415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
 416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
 417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
 418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
 419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   
 
         wbcc  rbcc  
 0     7300.0   6.4  
 1     7200.0   5.5  
 2     5500.0   5.7  
 3     820

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

The dataset is composed of different patients stats. Things like their blood pressure, red blood cell count, and sugar are recorded for each person. The data is composed of both numerical and categorical features. The labels are the numerical and categorical categories, and the features are the actual values themselves.

Categorical: specific gravity,sugar, albumin

Numerical: age, blood urea, packed cell volume

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

duplicates = sum(numerical_data.duplicated(keep= False, subset= ['unique_id']))
print(duplicates)

# Drop these duplicate rows

numerical_data = numerical_data.drop_duplicates(subset=['unique_id'], keep='last')
any = sum(numerical_data.duplicated(keep= False, subset= ['unique_id']))
print(any)


# Repeat the same for categorical dataset. Print the duplicate rows and drop them
categorical_data = categorical_data.drop_duplicates(subset=['unique_id'], keep='last')



59
0


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:Not all the unique ids are present in both datasets, because there are 10 less rows in the numerical dataset. In a combined table, if a id is missing from one of the datasets, then its valued would just be filled out with null or not present.

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(numerical_data, categorical_data, how='outer')
#Print the combined dataset
merged

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,...,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,...,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,...,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,...,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value
withMissing = len(merged)
mergedNoMissing = merged.dropna(axis=0,how='any')
noMissing = len(mergedNoMissing)
percent = noMissing/withMissing
# Print %
print(percent*100,"%")
# Drop these rows from the dataset
merged = mergedNoMissing
# Print the Dataset
merged

38.25 %


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,...,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,...,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
merged = merged.sort_values(by='Target')
merged = merged.reset_index()
# Print the dataset
merged

,index,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,...,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,...,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,...,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,105,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,...,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,108,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,...,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,110,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,...,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,171,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,43,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,169,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,54,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,...,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
merged = pd.get_dummies(merged, columns= categorical_columns)
# Print the dataset
merged 

,index,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,...,True,False,False,True,False,True,False,True,True,False
1,103,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,...,True,False,False,True,False,True,False,True,True,False
2,105,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,...,False,True,False,True,False,True,False,True,True,False
3,108,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,...,False,True,True,False,True,False,True,False,True,False
4,110,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,...,True,False,False,True,False,True,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,171,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,...,True,False,True,False,True,False,True,False,False,True
149,43,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,...,True,False,True,False,True,False,True,False,False,True
150,169,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,...,True,False,True,False,True,False,True,False,False,True
151,54,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

One column should be equivalent, because it conveys the same information. It would also be the same to have 5 set values, for categorial data. It should be the same because rather then just conveying the same inforamtion in 5 different columns, it would just give all of it in one column.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = merged[column].mean()
    std = merged[column].std()

    upper = mean +  3*std
    lower = mean - 3*std

    merged = merged[(merged[column] <= upper) & (merged[column] >= lower)]
# Print the dataset 
merged

,index,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,105,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,...,False,True,False,True,False,True,False,True,True,False
5,131,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,...,False,True,True,False,True,False,True,False,True,False
7,149,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,...,False,True,True,False,True,False,True,False,True,False
8,163,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,...,True,False,True,False,True,False,False,True,True,False
9,164,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,...,True,False,True,False,False,True,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,171,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,...,True,False,True,False,True,False,True,False,False,True
149,43,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,...,True,False,True,False,True,False,True,False,False,True
150,169,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,...,True,False,True,False,True,False,True,False,False,True
151,54,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
for column in numerical_columns:
    max = merged[column].max()
    min = merged[column].min()
    range = max-min
    
    merged[column] = (merged[column] - min)/range
# Print the dataset
merged 

,index,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
2,105,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,...,False,True,False,True,False,True,False,True,True,False
5,131,397388,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.206897,0.524752,...,False,True,True,False,True,False,True,False,True,False
7,149,438182,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,1.000000,0.277228,...,False,True,True,False,True,False,True,False,True,False
8,163,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,...,True,False,True,False,True,False,False,True,True,False
9,164,475200,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.586207,0.019802,...,True,False,True,False,False,True,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,171,486397,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.310345,0.683168,...,True,False,True,False,True,False,True,False,False,True
149,43,203466,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.482759,0.574257,...,True,False,True,False,True,False,True,False,False,True
150,169,482105,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.448276,0.722772,...,True,False,True,False,True,False,True,False,False,True
151,54,228849,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.206897,0.623762,...,True,False,True,False,True,False,True,False,False,True


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Unique id and index are not appropriate for modeling. They are only useful for people looking at the data and organizing the data int ables, however it does not actually hold any significance for the data.

In [11]:
#Remove that column
Columns = ['unique_id', 'index']
# Print the dataset
merged = merged.drop(columns=Columns)
merged = merged.reset_index(drop= True) 
merged

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.793103,0.000000,0.032258,0.395161,...,False,True,False,True,False,True,False,True,True,False
1,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.206897,0.524752,0.548387,0.443548,...,False,True,True,False,True,False,True,False,True,False
2,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,1.000000,0.277228,0.322581,0.233871,...,False,True,True,False,True,False,True,False,True,False
3,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.206897,0.059406,0.000000,0.653226,...,True,False,True,False,True,False,False,True,True,False
4,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.586207,0.019802,0.096774,0.258065,...,True,False,True,False,False,True,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.310345,0.683168,0.677419,0.096774,...,True,False,True,False,True,False,True,False,False,True
129,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.482759,0.574257,0.612903,0.290323,...,True,False,True,False,True,False,True,False,False,True
130,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.448276,0.722772,0.870968,0.274194,...,True,False,True,False,True,False,True,False,False,True
131,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.206897,0.623762,0.870968,0.193548,...,True,False,True,False,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
csv = merged.to_csv("chronic_kidney_disease.csv")
# Export the dataframe to a new json file
json = merged.to_json("chronic_kidney_disease.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [13]:
print(merged.head(15))

         age    bp       bgr        bu        sc       sod       pot  \
0   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
1   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
2   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
3   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
4   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
5   0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
6   0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
7   0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
8   0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
9   0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
10  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.689655   
11  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.862069   
12  0.878378  0.00  0.206009  0.751634  0.604938  0.533333  0.68

### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

You could manually check some of the data against the data table entries to see if it is doing the correct process. You could also give examples of how you want the output to look. 

|    age   |   bp  |   bgr   |   bu   |   sc   |   sod   |   pot   |    hemo   |    pcv   |    wbcc   | ... |   cad_no   |   cad_yes   |   appet_good   |   appet_poor   |   pe_no   |   pe_yes   |   ane_no   |   ane_yes   |   Target_ckd   |   Target_notckd   |
|----------|-------|---------|--------|--------|---------|---------|-----------|----------|-----------|-----|------------|--------------|-----------------|-----------------|-----------|------------|------------|--------------|----------------|-------------------|
| 0.743243 | 0.50  | 0.442060| 0.901961| 0.432099| 0.500000 | 0.793103 | 0.000000  | 0.032258 | 0.395161  | ... |   False    |    True      |      False      |      True        |   False    |   True     |   False    |    True      |      True      |       False         |
| 0.770270 | 1.00  | 0.901288| 0.163399| 0.345679| 0.766667 | 0.206897 | 0.524752  | 0.548387 | 0.443548  | ... |   False    |    True      |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.905405 | 0.50  | 0.785408| 0.862745| 0.518519| 0.600000 | 1.000000 | 0.277228  | 0.322581 | 0.233871  | ... |   False    |    True      |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.202703 | 0.75  | 0.158798| 0.196078| 0.160494| 0.166667 | 0.206897 | 0.059406  | 0.000000 | 0.653226  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   False    |    True      |      True      |       False         |
| 0.783784 | 1.00  | 0.399142| 0.287582| 0.839506| 0.666667 | 0.586207 | 0.019802  | 0.096774 | 0.258065  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.729730 | 0.00  | 0.935622| 0.169935| 0.160494| 0.333333 | 0.034483 | 0.019802  | 0.064516 | 0.879032  | ... |    True    |    False     |      False      |      True        |   True     |   False    |   False    |    True      |      True      |       False         |
| 0.675676 | 0.75  | 0.253219| 0.633987| 0.777778| 0.366667 | 0.655172 | 0.138614  | 0.193548 | 0.169355  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.851351 | 0.25  | 0.618026| 0.562092| 0.728395| 0.000000 | 0.344828 | 0.168317  | 0.161290 | 0.580645  | ... |   False    |    True      |      True       |      False       |   False    |   True     |   False    |    True      |      True      |       False         |
| 0.540541 | 0.00  | 0.399142| 0.535948| 0.358025| 0.700000 | 0.379310 | 0.207921  | 0.161290 | 0.830645  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.756757 | 0.25  | 0.223176| 0.209150| 0.160494| 0.533333 | 0.620690 | 0.485149  | 0.516129 | 0.290323  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.662162 | 0.50  | 0.618026| 0.411765| 0.432099| 0.566667 | 0.689655 | 0.316832  | 0.354839 | 0.250000  | ... |    True    |    False     |      True       |      False       |   True     |   False    |   False    |    True      |      True      |       False         |
| 0.783784 | 0.00  | 0.725322| 0.313725| 0.481481| 0.566667 | 0.862069 | 0.178218  | 0.193548 | 0.258065  | ... |    True    |    False     |      False      |      True        |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.878378 | 0.00  | 0.206009| 0.751634| 0.604938| 0.533333 | 0.689655 | 0.366337  | 0.387097 | 0.879032  | ... |    True    |    False     |      False      |      True        |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.878378 | 0.25  | 0.639485| 0.470588| 0.395062| 0.433333 | 0.517241 | 0.267327  | 0.322581 | 0.104839  | ... |   False    |    True      |      True       |      False       |   True     |   False    |   True     |    False     |      True      |       False         |
| 0.716216 | 0.50  | 1.000000| 0.163399| 0.111111| 0.066667 | 0.206897 | 0.267327  | 0.387097 | 0.532258  | ... |    True    |    False     |      False      |      True        |   True     |   False    |   True     |    False     |      True      |       False         |


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [14]:
# Code to rename all the columns in the dataset
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Renaming columns
merged.rename(columns=rename_dict, inplace=True)

# Display the renamed DataFrame
print("Renamed DataFrame:")
print(merged)

Renamed DataFrame:
          age  blood pressure  blood glucose random  blood urea  \
0    0.743243            0.50              0.442060    0.901961   
1    0.770270            1.00              0.901288    0.163399   
2    0.905405            0.50              0.785408    0.862745   
3    0.202703            0.75              0.158798    0.196078   
4    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
128  0.189189            0.25              0.227468    0.222222   
129  0.216216            0.00              0.115880    0.052288   
130  0.513514            0.00              0.107296    0.235294   
131  0.702703            0.50              0.128755    0.261438   
132  0.689189            0.00              0.150215    0.254902   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0            0.432099  0.500000   0.793103    0.000000            0.032258   
1            0.34567

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [15]:
# Converted SQL to Pandas code

# Grouping by 'Target' and counting occurrences
result = merged.groupby('Target_ckd').size().reset_index(name='count')

# Displaying the result
print(result)


   Target_ckd  count
0       False    111
1        True     22


*Describe what the above code does here*

The query goes through all the data entries, and identifies the possible values of Target_ckd and how many occurrences there are of each.